In [771]:
# импорт библиотеки pandas
import pandas as pd
import numpy as np

In [772]:
#чтение датасета с помощью библиотеки pandas
df = pd.read_csv('5music.csv')

## Вывести первые 20 строк с помощью head:

In [773]:
df.head(20)

userID                                          trackname  \
0   FFB692EC                                  Kamigata To Boots   
1   55204538                        Delayed Because of Accident   
2     20EC38                                  Funiculì funiculà   
3   A3DD03C9                              Dragons in the Sunset   
4   E2DC1FAE                                        Soul People   
5   842029A1                                          Преданная   
6   4CB90AA5                                               True   
7   F03E1C1F                                   Feeling This Way   
8   8FA1D3BE                           И вновь продолжается бой   
9   E772D5C0                                          Pessimist   
10  BC5A3A29                                       Gool la Mita   
11  8B5192C0  Is There Anybody Out There? (Panoramic Paralys...   
12  FF3FD2BD                                Мина син генэ кирэк   
13  CC782B0F                               After School Special   
14  94EB25C2                         Make Love Whenever You Can   
15  E3C5756F                                                NaN   
16  81D05C7D                                     SLAVES OF FEAR   
17  39DE290E                                           Веснянка   
18  58AE138A                                         Pat-a-Cake   
19  772F5B59                                          Sweetback   

              Artistname        genre            City        time    Dayweek;;  
0       The Mass Missile         rock  Saint-Petersburg  20:28:33  Wednesday;;  
1       Andreas Rönnberg         rock            Moscow  14:07:09     Friday;;  
2            Mario Lanza          pop  Saint-Petersburg  20:58:07  Wednesday;;  
3             Fire + Ice         folk  Saint-Petersburg  08:37:09     Monday;;  
4             Space Echo        dance            Moscow  08:34:34     Monday;;  
5              IMPERVTOR       rusrap  Saint-Petersburg  13:09:41     Friday;;  
6           Roman Messer        dance            Moscow  13:00:07  Wednesday;;  
7        Polina Griffith        dance            Moscow  20:47:49  Wednesday;;  
8                    NaN       ruspop            Moscow  09:17:40     Friday;;  
9                    NaN        dance  Saint-Petersburg  21:20:49  Wednesday;;  
10   Shireen Abdul Wahab        world            Moscow  14:08:42     Monday;;  
11  Pink Floyd Floydhead   electronic            Moscow  13:47:49     Monday;;  
12        Ильдар Хакимов          pop            Moscow  09:19:49     Monday;;  
13  Detroit Grand Pubahs        dance            Moscow  20:04:12     Friday;;  
14             Arabesque          pop  Saint-Petersburg  13:22:08  Wednesday;;  
15                   NaN          NaN            Moscow  09:24:51     Monday;;  
16                HEALTH  alternative  Saint-Petersburg  20:54:48     Monday;;  
17         Веданъ Колодъ         folk  Saint-Petersburg  14:36:47     Monday;;  
18     Mother Goose Club     children  Saint-Petersburg  20:04:56     Monday;;  
19    The Fabulous Three          rnb            Moscow  13:56:42  Wednesday;;

## Выполнить обзор данных – опишите колонки и вашу предметную область: 

### Предметная область - музыка. 
Мы можем узнать название песни, исполнителя, жанр, идентификатор пользователя. 

Насчет времени сложно сказать, что это продолжительность песни, возможно, это время проигрывания, например, на радио, время, в которое пользователь начал прослушаивать песню. 

Можно предположить, что идет сравнение пользователей в двух городах СПб и Мск, а также сравнение жанров по дням недели (какой жанр в какой день предпочтительнее).

Датасет содержит следующие колонки:
- userID - идентификатор пользователя,
- trackname - название песни,
- Artistname - исполнитель,
- genre - жанр,
- City - город,
- time - время,
- DayWeek - день недели.

In [774]:
#оценка данных
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65215 entries, 0 to 65214
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0     userID    65215 non-null  object
 1   trackname   63256 non-null  object
 2   Artistname  57330 non-null  object
 3   genre       63292 non-null  object
 4     City      64487 non-null  object
 5   time        64487 non-null  object
 6   Dayweek;;   64485 non-null  object
dtypes: object(7)
memory usage: 3.5+ MB


### С помощью метода .info оцените данные. 
### Этот метод возвращает название столбцов, типы данных, количество ненулевых объектов каждом столбце.

- В датасете содержится 65215 строк, 7 столбцов. 
- У всех полей тип данных - object. 
- Присутствуют некорректные названия столбцов. 
- Количество значений в столбцах различается. Значит, в данных есть пропущенные значения.


In [775]:
#вывод названий столбцов
df.columns

Index(['  userID', 'trackname', 'Artistname', 'genre', '  City  ', 'time',
       'Dayweek;;'],
      dtype='object')

### Проблемы с названиями столбцов:
- присутсвие пробелов в названии (столбцы: city, userID),
- cтрочные буквы сочетаются с прописными (Artistname, City, Dayweek),
- лишние символы в названии столбца Dayweek,
- отсутствие "змеиного стиля" (присутсвие нижнего подчеркивания) при названии из нескольких слов (userID, trackname, Artistname, Dayweek).


In [776]:
#приводим все наименования к нижнему регистру
df.columns = df.columns.str.lower()
df.columns

Index(['  userid', 'trackname', 'artistname', 'genre', '  city  ', 'time',
       'dayweek;;'],
      dtype='object')

In [777]:
#переименования столбцов
df=df.rename(columns={'  userid': 'user_id', 'trackname':
'track_name', 'artistname': 'artist_name', '  city  ': 'city', 'dayweek;;': 'day_week'})
df.columns


Index(['user_id', 'track_name', 'artist_name', 'genre', 'city', 'time',
       'day_week'],
      dtype='object')

## Поиск пропусков данных

In [778]:
# подсчёт пропусков
df.isna().sum()

user_id           0
track_name     1959
artist_name    7885
genre          1923
city            728
time            728
day_week        730
dtype: int64

In [779]:
# удалим строки в столбце artist_name, так как этим данные не самые важные, у песни может и не быть исполнителя
df = df.dropna(subset=['artist_name'])
df.isna().sum()


user_id          0
track_name       0
artist_name      0
genre          364
city             0
time             0
day_week         2
dtype: int64

In [780]:
# в столбцах: genre, city, day_week заменим данные на "неизвестно"
columns_to_replace = ['track_name', 'genre', 'city', 'day_week']
for col_name in columns_to_replace:
    df[col_name] = df[col_name].fillna('unknown')

In [781]:
# в столбцe: time заменим данные на "0" и проверим остались ли у нас пропуски
df['time'] = df['time'].fillna(0)
df.isna().sum()

user_id        0
track_name     0
artist_name    0
genre          0
city           0
time           0
day_week       0
dtype: int64

От пропусков избавивились.
## Поиск дубликатов данных

In [782]:
# подсчёт явных дубликатов
df.duplicated().sum()

3365

In [783]:
# удаление явных дубликатов (с удалением старых индексов и формированием новых)
df=df.drop_duplicates().reset_index(drop = True)

In [784]:
# проверка на отсутствие дубликатов
df.duplicated().sum()

0

In [785]:
# просмотр уникальных названий жанров (поиск неявных дубликатов)
df['genre'].sort_values().unique()

array(['Moscow', 'acid', 'acoustic', 'action', 'adult', 'africa',
       'afrikaans', 'alternative', 'alternativepunk', 'ambient',
       'americana', 'animated', 'anime', 'arabesk', 'arabic', 'arena',
       'argentinetango', 'art', 'audiobook', 'avantgarde', 'axé', 'baile',
       'balkan', 'beats', 'bigroom', 'black', 'bluegrass', 'blues',
       'bollywood', 'bossa', 'brazilian', 'breakbeat', 'breaks',
       'broadway', 'cantautori', 'cantopop', 'canzone', 'caribbean',
       'caucasian', 'celtic', 'chamber', 'chanson', 'children', 'chill',
       'chinese', 'choral', 'christian', 'christmas', 'classical',
       'classicmetal', 'club', 'colombian', 'comedy', 'conjazz',
       'contemporary', 'country', 'cuban', 'dance', 'dancehall',
       'dancepop', 'dark', 'death', 'deep', 'deutschrock', 'deutschspr',
       'disco', 'dnb', 'documentary', 'downbeat', 'downtempo', 'drum',
       'dub', 'dubstep', 'eastern', 'easy', 'electronic', 'electropop',
       'emo', 'entehno', 'epicmetal

В массиве названий жанров присутствует повтор или альтернативное название: hiphop.

In [786]:
# функция для замены неявных дубликатов
def replace_wrong_genre (wrong_genres,correct_genre):
    for wrong_genre in wrong_genres:
         df['genre'] = df['genre'].replace(wrong_genre, correct_genre)

In [787]:
# устранение неявных дубликатов
duplicates = ['hip','hop','hip-hop']
genre = 'hiphop' # правильное имя
replace_wrong_genre(duplicates, genre)# вызов функции


duplicates2 = ['электроника']
genre2 = 'electronic' # правильное имя
replace_wrong_genre(duplicates2, genre2)# вызов функции

In [788]:
df = df.drop(np.where(df['genre'] == 'Moscow')[0])

In [789]:
# проверка на неявные дубликаты
print (df['genre'].sort_values().unique())

['acid' 'acoustic' 'action' 'adult' 'africa' 'afrikaans' 'alternative'
 'alternativepunk' 'ambient' 'americana' 'animated' 'anime' 'arabesk'
 'arabic' 'arena' 'argentinetango' 'art' 'audiobook' 'avantgarde' 'axé'
 'baile' 'balkan' 'beats' 'bigroom' 'black' 'bluegrass' 'blues'
 'bollywood' 'bossa' 'brazilian' 'breakbeat' 'breaks' 'broadway'
 'cantautori' 'cantopop' 'canzone' 'caribbean' 'caucasian' 'celtic'
 'chamber' 'chanson' 'children' 'chill' 'chinese' 'choral' 'christian'
 'christmas' 'classical' 'classicmetal' 'club' 'colombian' 'comedy'
 'conjazz' 'contemporary' 'country' 'cuban' 'dance' 'dancehall' 'dancepop'
 'dark' 'death' 'deep' 'deutschrock' 'deutschspr' 'disco' 'dnb'
 'documentary' 'downbeat' 'downtempo' 'drum' 'dub' 'dubstep' 'eastern'
 'easy' 'electronic' 'electropop' 'emo' 'entehno' 'epicmetal' 'estrada'
 'ethnic' 'eurofolk' 'european' 'experimental' 'extrememetal' 'fado'
 'fairytail' 'film' 'fitness' 'flamenco' 'folk' 'folklore' 'folkmetal'
 'folkrock' 'folktronica' 'fo

## Проверьте все ли типы данных соответствуют действительности. 


In [790]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 53963 entries, 0 to 53964
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      53963 non-null  object
 1   track_name   53963 non-null  object
 2   artist_name  53963 non-null  object
 3   genre        53963 non-null  object
 4   city         53963 non-null  object
 5   time         53963 non-null  object
 6   day_week     53963 non-null  object
dtypes: object(7)
memory usage: 3.3+ MB


У всех один тип данных - object.
Заменим все текстовые поля (city, track_name, artist_name, genre, day_week) на string.

In [791]:
df['track_name'] = df['track_name'].astype('string')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 53963 entries, 0 to 53964
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      53963 non-null  object
 1   track_name   53963 non-null  string
 2   artist_name  53963 non-null  object
 3   genre        53963 non-null  object
 4   city         53963 non-null  object
 5   time         53963 non-null  object
 6   day_week     53963 non-null  object
dtypes: object(6), string(1)
memory usage: 3.3+ MB


In [792]:
df['day_week'] = df['day_week'].astype('string')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 53963 entries, 0 to 53964
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      53963 non-null  object
 1   track_name   53963 non-null  string
 2   artist_name  53963 non-null  object
 3   genre        53963 non-null  object
 4   city         53963 non-null  object
 5   time         53963 non-null  object
 6   day_week     53963 non-null  string
dtypes: object(5), string(2)
memory usage: 3.3+ MB


In [793]:
df['genre'] = df['genre'].astype('string')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 53963 entries, 0 to 53964
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      53963 non-null  object
 1   track_name   53963 non-null  string
 2   artist_name  53963 non-null  object
 3   genre        53963 non-null  string
 4   city         53963 non-null  object
 5   time         53963 non-null  object
 6   day_week     53963 non-null  string
dtypes: object(4), string(3)
memory usage: 3.3+ MB


In [794]:
df['city'] = df['city'].astype('string')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 53963 entries, 0 to 53964
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      53963 non-null  object
 1   track_name   53963 non-null  string
 2   artist_name  53963 non-null  object
 3   genre        53963 non-null  string
 4   city         53963 non-null  string
 5   time         53963 non-null  object
 6   day_week     53963 non-null  string
dtypes: object(3), string(4)
memory usage: 3.3+ MB


In [795]:
df['artist_name'] = df['artist_name'].astype('string')
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 53963 entries, 0 to 53964
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_id      53963 non-null  object
 1   track_name   53963 non-null  string
 2   artist_name  53963 non-null  string
 3   genre        53963 non-null  string
 4   city         53963 non-null  string
 5   time         53963 non-null  object
 6   day_week     53963 non-null  string
dtypes: object(2), string(5)
memory usage: 3.3+ MB


## Придумайте, какую сводную таблицу можно создать по вашим данным и создайте сводную таблицу с помощью pivot_table.

Составим свобдную таблицу, где увидим сколько пользователей слушает определенный жанр и в каком городе.

Для наглядности выведем не все жанры , а 50.


In [805]:
gg = pd.pivot_table(df,
               index=['genre'],
               columns=['city'],
               values=['time'],
               aggfunc=['count'],
               fill_value=0,
               )
gg.head(50)
               #margins=True

count                 
                            time                 
city            Moscow           Saint-Petersburg
genre                                            
acid                           1                0
acoustic                       1                1
action                         3                0
adult                         16                8
africa                        10                4
afrikaans                      4                0
alternative                 1332              622
alternativepunk                0                1
ambient                      169               59
americana                      7                1
animated                       2                0
anime                         72               28
arabesk                        7                2
arabic                         1                1
arena                          1                0
argentinetango                18                6
art                            2                0
audiobook                      6                1
avantgarde                    42               13
axé                            1                0
baile                          0                1
balkan                         7                4
beats                         13                5
bigroom                        9                3
black                         18                3
bluegrass                      4                2
blues                        352              110
bollywood                      2                2
bossa                          1                0
brazilian                     18                4
breakbeat                      4                3
breaks                         4                2
broadway                      10                4
cantautori                     2                0
cantopop                       2                0
canzone                        2                1
caribbean                      3                0
caucasian                     72               26
celtic                        14                1
chamber                        3                2
chanson                       53               14
children                     353              117
chill                         60               35
chinese                       11                5
choral                         3                0
christian                     59               23
christmas                     13                6
classical                   1200              482
classicmetal                  42               17
club                           9                0